cpp/ql/src/Critical/UseAfterFree.ql

```diff
@@ -11,56 +11,26 @@
  */
 
 import cpp
-import semmle.code.cpp.controlflow.StackVariableReachability
-
-/** `e` is an expression that frees the memory pointed to by `v`. */
-predicate isFreeExpr(Expr e, StackVariable v) {
-  exists(VariableAccess va | va.getTarget() = v |
-    exists(FunctionCall fc | fc = e |
-      fc.getTarget().hasGlobalOrStdName("free") and
-      va = fc.getArgument(0)
-    )
-    or
-    e.(DeleteExpr).getExpr() = va
-    or
-    e.(DeleteArrayExpr).getExpr() = va
-  )
-}
-
-/** `e` is an expression that (may) dereference `v`. */
-predicate isDerefExpr(Expr e, StackVariable v) {
-  v.getAnAccess() = e and dereferenced(e)
-  or
-  isDerefByCallExpr(_, _, e, v)
+import semmle.code.cpp.dataflow.new.DataFlow
+import FlowAfterFree as Flow
+
+module UseAfterFreeFlow = Flow::FlowFromFree<useExpr/2>;
+
+predicate useExpr(DataFlow::Node dfe, Expr e) {
+  e = dfe.asExpr() and
+  (
+    e = any(ReturnStmt rs).getExpr() or
+    e = any(FunctionCall fc).getAChild() or
```

Comment:
The first case. The intention is flag passing a pointer pointing to de-allocated memory to a function, indeed

We could extend the analysis to be interprocedural, but I would argue that passing an invalid pointer to any function should be considered harmful, and  I have not seen any FP's for that case. 
The only point it would make sense is if the callee just ignores the argument. 

It should be noted that we block flow though `CallByReference`.

Reply:
Followup after a sync talk. This would avoid an FP in 
```
void *a = ...
free(a);
printf("%p", a);
```